RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporterWithVeto.cc

```diff
@@ -0,0 +1,211 @@
+  auto btrack = tracks->cbegin();
+  auto etrack = tracks->cend();
+  for( auto track = btrack;  track != etrack; ++track) {
+    auto tk_elem = std::find_if(btk_elems,TKs_end,
+				[&](const ElementType& a){
+				  return ( a->trackRef() == 
+					   track->trackRef() );
+				});
+    if( tk_elem != TKs_end ) {
+      mask[std::distance(tracks->cbegin(),track)] = false;
+      // check and update if this track is a muon
+      const int muId = muAssocToTrack( (*tk_elem)->trackRef(), muons );
+      if( muId != -1 ) {
+	muonref= reco::MuonRef( muons, muId );
+	if( PFMuonAlgo::isLooseMuon(muonref) || PFMuonAlgo::isMuon(muonref) ) {
+	  static_cast<reco::PFBlockElementTrack*>(tk_elem->get())->setMuonRef(muonref);
+	}
+      }
+    }    
+  }
+  // now we actually insert tracks, again tagging muons along the way
+  reco::PFRecTrackRef pftrackref;  
+  reco::PFBlockElementTrack* trkElem = NULL;
+  for( auto track = btrack;  track != etrack; ++track) {
+    const unsigned idx = std::distance(btrack,track);
+    // since we already set muon refs in the previously imported tracks,
+    // here we can skip everything that is already imported 
+    if( !mask[idx] ) continue; 
+    muonref = reco::MuonRef();
+    pftrackref = reco::PFRecTrackRef(tracks,idx);    
+    if( vetoed.count(pftrackref->trackRef().key()) ) continue;
+    // Get the eventual muon associated to this track
+    const int muId = muAssocToTrack( pftrackref->trackRef(), muons );
+    bool thisIsAPotentialMuon = false;
+    if( muId != -1 ) {
+      muonref= reco::MuonRef( muons, muId );
+      thisIsAPotentialMuon = ( (pfmu_->hasValidTrack(muonref,true)&&PFMuonAlgo::isLooseMuon(muonref)) || 
+			       (pfmu_->hasValidTrack(muonref,false)&&PFMuonAlgo::isMuon(muonref)));
+    }
+    if(thisIsAPotentialMuon || goodPtResolution( pftrackref->trackRef() ) ) {
+      trkElem = new reco::PFBlockElementTrack( pftrackref );
+      if (thisIsAPotentialMuon && _debug) {
+	std::cout << "Potential Muon P " <<  pftrackref->trackRef()->p() 
+		  << " pt " << pftrackref->trackRef()->p() << std::endl; 
+      }
+      if( muId != -1 ) trkElem->setMuonRef(muonref);
+      elems.emplace_back(trkElem);
+    }
+  }
+  elems.shrink_to_fit();
+}
+
+bool GeneralTracksImporterWithVeto::
+goodPtResolution( const reco::TrackRef& trackref) const {
+  //recheck that the track is high purity!
+  if (!trackref->quality(reco::TrackBase::highPurity))
+    return false;
+    
+
+  const double P = trackref->p();
+  const double Pt = trackref->pt();
+  const double DPt = trackref->ptError();
+  const unsigned int NHit = 
+    trackref->hitPattern().trackerLayersWithMeasurement();
+  const unsigned int NLostHit = 
+    trackref->hitPattern().trackerLayersWithoutMeasurement(reco::HitPattern::TRACK_HITS);
+  const unsigned int LostHits = trackref->numberOfLostHits();
+  const double sigmaHad = sqrt(1.20*1.20/P+0.06*0.06) / (1.+LostHits);
+
+  // Protection against 0 momentum tracks
+  if ( P < 0.05 ) return false;
```

Comment:
Yes, trajectory states that have 0 pT cannot be in the tracker volume and high purity (pointing at beamspot).